### PR TITLE
message/messaging_service: guard adding maintenance tenant under cluster feature

### PR DESCRIPTION
**Pull request by Jadw1**
In https://github.com/scylladb/scylladb/pull/18729, we introduced a new statement tenant `$maintenance`, but the change wasn't protected by any cluster feature.
This wasn't a problem for OSS, since unknown isolation cookie just uses default scheduling group. However, in enterprise that leads to creating a service level on not-upgraded nodes, which may end up in an error if user create maximum number of service levels.

This patch adds a cluster feature to guard adding the new tenant. It's done in the way to handle two upgrade scenarios:
- version without `$maintenance` tenant -> version with `$maintenance` tenant guarded by a feature
- version with `$maintenance` tenant but not guarded by a feature -> version with `$maintenance` tenant guarded by a feature

The PR adds `enabled` flag to statement tenants.
This way, when the tenant is disabled, it cannot be used to create a connection, but it can be used to accept an incoming connection.
The `$maintenance` tenant is added to the config as disabled and it gets enabled once the corresponding feature is enabled.

Fixes scylladb/scylladb#20070 
Refs scylladb/scylla-enterprise#4403

### CODE REVIEW

**Comment by Jadw1** [comment]
@denesb should I backport this to: 5.2, 5.4, 6.0 and 6.1? https://github.com/scylladb/scylladb/pull/18729 was backported to 5.2, 5.4, 6.0

**Comment by mykaul** [comment]
> @denesb should I backport this to: 5.2, 5.4, 6.0 and 6.1? #18729 was backported to 5.2, 5.4, 6.0

We only support 5.4 and 6.0

File: message/messaging_service.cc
Comment (denesb) [inline message/messaging_service.cc:712]: Why wrap into a function? Wouldn't the old fall-back of `return idx` work just as fine?

File: message/messaging_service.cc
Comment (denesb) [inline message/messaging_service.cc:452]: Can we do this in main somehow? Then we don't need the previous patch and the hard-coding happens at the place where we supply the had-coded config anyway?
We just need a mutable `get_tenant_config()` for this, but I think that is the lesser of evils.

File: message/messaging_service.cc
Comment (Jadw1) [reply message/messaging_service.cc:712]: It still works the same but the tenant is disabled, then the function is run once again looking for default tenant. That's I've extracted it to a function.

**Comment by avikivity** [comment]
I think a better fix is to catch \$.* on enterprise and map it to the default level. This allows us to have reasonable fallbacks in the future too.

Also, b2a500a9a14eb1debef1271336d1d453420eb0c5 was backported and is already in the field, so we can't rely on the new cluster feature for those versions.

**Comment by Jadw1** [comment]
> I think a better fix is to catch $.* on enterprise and map it to the default level. This allows us to have reasonable fallbacks in the future too.

I agree with that but user can create a service level with name starting with `$`. Would it be okay to forbid this kind of names in service levels?

File: message/messaging_service.cc
Comment (piodul) [reply message/messaging_service.cc:712]: I don't think it will work the same. What I think that Botond imagined is just changing

```c++
if (_connection_index_for_tenant[i].sched_group == curr_sched_group) {
```

to

```c++
if (_connection_index_for_tenant[i].sched_group == curr_sched_group && _connection_index_for_tenant[i].enabled) {
```

and letting `return idx` handle it if the tenant is not enabled. If the maintenance tenant is not enabled, the `return idx` will return a value from range `[0, PER_TENANT_CONNECTION_COUNT)` - or, in other terms, the one that would correspond to the tenant with idx `0` (`"$user"`).

In the current solution, we will look for a connection corresponding to the tenant assigned to the default scheduling group (`"$system"`). This tenant has index **1**, not 0, so it will return something from range `[PER_TENANT_CONNECTION_COUNT, 2*PER_TENANT_CONNECTION_COUNT)`.

@denesb which tenant is preferred in this case, `$user` or `$system`?

**Comment by piodul** [comment]
> > > I think a better fix is to catch $.* on enterprise and map it to the default level. This allows us to have reasonable fallbacks in the future too.
> > 
> > 
> > I agree with that but user can create a service level with name starting with `$`. Would it be okay to forbid this kind of names in service levels?
> 
> Yes, "$" is reserved. Maybe not formally, but we can formalize it by rejecting service levels with that character.
> 
> We also used it in CDC column names.

I think we can avoid having to restrict valid service level names by introducing a prefix that we would add when putting the service level names in the isolation cookie string. So, if it starts with `$` then we know it's a tenant, if it starts with `#` (or any other more sensible string) then it's a service level.

**Comment by Jadw1** [comment]
> > > > I think a better fix is to catch $.* on enterprise and map it to the default level. This allows us to have reasonable fallbacks in the future too.
> > > 
> > > 
> > > I agree with that but user can create a service level with name starting with `$`. Would it be okay to forbid this kind of names in service levels?
> > 
> > 
> > Yes, "$" is reserved. Maybe not formally, but we can formalize it by rejecting service levels with that character.
> > We also used it in CDC column names.
> 
> I think we can avoid having to restrict valid service level names by introducing a prefix that we would add when putting the service level names in the isolation cookie string. So, if it starts with `$` then we know it's a tenant, if it starts with `#` (or any other more sensible string) then it's a service level.

But the solution with 2 prefixes would require a cluster feature, so it won't help with versions where https://github.com/scylladb/scylladb/commit/b2a500a9a14eb1debef1271336d1d453420eb0c5 was already backported.

However I think we can use default scheduling group for unknown isolation cookie starting with `$`  unconditionally (without any cluster feature) and add a prefix to service level isolation cookie when a cluster feature is enabled.

**Comment by avikivity** [comment]
> > > > I think a better fix is to catch $.* on enterprise and map it to the default level. This allows us to have reasonable fallbacks in the future too.
> > > 
> > > 
> > > I agree with that but user can create a service level with name starting with `$`. Would it be okay to forbid this kind of names in service levels?
> > 
> > 
> > Yes, "$" is reserved. Maybe not formally, but we can formalize it by rejecting service levels with that character.
> > We also used it in CDC column names.
> 
> I think we can avoid having to restrict valid service level names by introducing a prefix that we would add when putting the service level names in the isolation cookie string. So, if it starts with `$` then we know it's a tenant, if it starts with `#` (or any other more sensible string) then it's a service level.

We can't change the encoding now (well we can, but with some effort).

I think we can take the risk that if someone will name their service level "$system" they will get surprising results.

**Comment by mykaul** [comment]
> I think we can take the risk that if someone will name their service level "$system" they will get surprising results.

https://xkcd.com/327/

File: message/messaging_service.cc
Comment (Jadw1) [reply message/messaging_service.cc]: Removed the wrapper function. If no tenant is found, then `$user` is used.

When this PR will be merged to enterprise, this code has to be changed to use default tenant when tenant for current scheduling group is disabled and to not fall into [service levels tenant](https://github.com/scylladb/scylla-enterprise/blob/enterprise/message/messaging_service.cc#L736-L761).

**Comment by Jadw1** [comment]
Changelog:
- applied change to [the comment](https://github.com/scylladb/scylladb/pull/19802#discussion_r1698640660)
- a few tests are still failing, I'm still looking at them but I've pushed the code for review

File: message/messaging_service.cc
Comment (denesb) [reply message/messaging_service.cc]: > @denesb which tenant is preferred in this case, $user or $system?

We have a "default" tenant, the first tenant in the tenant configuration. That one should be used, to allow the default tenant to be selected by the configuration, instead of hard-coding it.

**Comment by avikivity** [comment]
I got lost in the discussion. Why not detect $.* as internal tenants?

**Comment by piodul** [comment]
> I got lost in the discussion. Why not detect $.* as internal tenants?

@avikivity https://github.com/scylladb/scylladb/pull/19802#issuecomment-2250222877

Citing my previous comment:

> > I think we can take the risk that if someone will name their service level "$system" they will get surprising results.
>
> Well, we are already taking this risk right now, ($system, $user and $maintenance). Capturing all names that match the regex `\$.*` would increase this risk even further.

**Comment by Jadw1** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10714/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10714/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2877/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10714/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10899/)
> 
> #### Failed Tests (1/31568):
> * [test_increment_decrement_counters_in_threads_nodes_restarted](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10714/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_increment_decrement_counters_in_threads_nodes_restarted) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_increment_decrement_counters_in_threads_nodes_restarted)

The failed test seems unrelated to this PR: https://github.com/scylladb/scylla-dtest/issues/3686

**Comment by Jadw1** [comment]
@piodul ping

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
@Jadw1 One thing before merge: for the sake of backports - for which CI will fail if you won't do the following - please create an issue in OSS describing this problem and change the description of this PR so that it "Fixes" that issue.

**Comment by Jadw1** [comment]
> @Jadw1 One thing before merge: for the sake of backports - for which CI will fail if you won't do the following - please create an issue in OSS describing this problem and change the description of this PR so that it "Fixes" that issue.

done

**Comment by mykaul** [comment]
@scylladb/scylla-maint - please review for merge.
@Jadw1 - please resolve fixed comments, so it'll be easier to re-review.

**Comment by Jadw1** [comment]
> @scylladb/scylla-maint - please review for merge. @Jadw1 - please resolve fixed comments, so it'll be easier to re-review.

After discussing [Avi's comment](https://github.com/scylladb/scylladb/pull/19802#issuecomment-2273910311) with @piodul, we decided to follow Avi's sugggestion and fix by using default scheduling group for every unknown tenant starting with `$` instead of doing this PR.

So pleas don't merge this, also you don't have to review. I'll post the new PR right after I fix https://github.com/scylladb/scylla-enterprise/issues/4567 and https://github.com/scylladb/scylla-enterprise/issues/4566. I'll close this PR after I submit the new one.

**Comment by susiesolis** [comment]
> > @scylladb/scylla-maint - please review for merge. @Jadw1 - please resolve fixed comments, so it'll be easier to re-review.
> 
> After discussing [Avi's comment](https://github.com/scylladb/scylladb/pull/19802#issuecomment-2273910311) with @piodul, we decided to follow Avi's sugggestion and fix by using default scheduling group for every unknown tenant starting with `$` instead of doing this PR.
> 
> So pleas don't merge this, also you don't have to review. I'll post the new PR right after I fix [scylladb/scylla-enterprise#4567](https://github.com/scylladb/scylla-enterprise/issues/4567) and [scylladb/scylla-enterprise#4566](https://github.com/scylladb/scylla-enterprise/issues/4566). I'll close this PR after I submit the new one.

Is there an ETA we can give the customer? They are asking.

**Comment by pidiaquez** [comment]
Will this fix be added to next release or custom patch will be provided for the affected customer?

**Comment by Jadw1** [comment]
Opened https://github.com/scylladb/scylladb/pull/20122

**Comment by Jadw1** [comment]
https://github.com/scylladb/scylladb/pull/20122 got merged, closing this PR without merging

**Comment by Jadw1** [comment]
Reopening the issue.

Fix done in https://github.com/scylladb/scylladb/pull/20122 is not enough because it doesn't handle the case when a node with `$maintenance` tenant tries create a RPC connection to not-yet-upgraded node.

This PR handles it by adding a cluster feature and disabling `$maintenance` use to create new connections until the cluster feature is off.

There is no collision with https://github.com/scylladb/scylladb/pull/20122.

**Comment by Jadw1** [comment]
Changelog:
- rebased on top of latest master

**Comment by piodul** [comment]
> * [test_concurrent_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11639/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest_with_topology_changes___test_concurrent_restart) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_restart)

This is most likely https://github.com/scylladb/scylla-dtest/issues/4303

I have no clue about the other two, I didn't see corresponding issues in scylladb or scylla-dtest repos.

- There was a segfault in [test_initial_token.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11639/testReport/junit/%28root%29/test_initial_token/Tests___Unit_Tests___test_initial_token_dev_2) (scylla-2967.log, `raft_topology - raft_state_monitor_fiber aborted with raft::stopped_error (Raft instance is stopped, reason: "raft group0 is stopped")`)
- I don't see how [test_data_distribution_balance[4-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11639/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_4_TimeWindowCompactionStrategy_)  could be related, but I saw that the test was recently modified in https://github.com/scylladb/scylla-dtest/pull/4893. Maybe it became flaky?

@Jadw1 please decode the backtrace from test_initial_token and look into it. As for test_data_distribution_balance, if you don't see how it is related then please open an issue about the failure.

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Queued

**Comment by piodul** [comment]
I removed the backport/5.4 and backport/6.0 labels (6.0 will soon be obsolete, and 5.4 already is).

This will have to be backported to 6.2, but I don't see the corresponding backport label yet.

**Pull request closed without merging**
